cl/citations/tasks.py

```diff
@@ -126,32 +133,45 @@ def find_citations_for_opinion_by_pks(self, opinion_pks, index=True):
         # values = number of times that opinion is cited
         grouped_matches = Counter(citation_matches)
 
+        # Increase citation count for each matched cluster if it hasn't
+        # already been cited by this opinion. First, calculate a list
+        # of the ID of every opinion that will need need updating.
+        all_cited_opinions = opinion.opinions_cited.all().values_list(
+            "pk", flat=True
+        )
+        opinion_ids_to_update = set()
         for matched_opinion in grouped_matches:
-            # Increase citation count for matched cluster if it hasn't
-            # already been cited by this opinion.
-            if matched_opinion not in opinion.opinions_cited.all():
-                matched_opinion.cluster.citation_count += 1
-                matched_opinion.cluster.save(index=index)
-
-        # Only update things if we found citations
-        if citations:
-            opinion.html_with_citations = create_cited_html(opinion, citations)
-
-            # Nuke existing citations
-            OpinionsCited.objects.filter(citing_opinion_id=opinion.pk).delete()
-
-            # Create the new ones.
-            OpinionsCited.objects.bulk_create(
-                [
-                    OpinionsCited(
-                        citing_opinion_id=opinion.pk,
-                        cited_opinion_id=matched_opinion.pk,
-                        depth=grouped_matches[matched_opinion],
-                    )
-                    for matched_opinion in grouped_matches
-                ]
+            if matched_opinion.pk not in all_cited_opinions:
+                opinion_ids_to_update.add(matched_opinion.pk)
+
+        # Then, increment the citation_count fields for those matched clusters
+        # all at once. Trigger a single Solr update as well, if required.
+        OpinionCluster.objects.filter(
+            sub_opinions__pk__in=opinion_ids_to_update
+        ).update(citation_count=F("citation_count") + 1)
+        if index:
+            add_items_to_solr.delay(
+                opinion_ids_to_update, "search.OpinionCluster",
             )
 
+        # Generate the opinion's new HTML (with inline citation links)
+        opinion.html_with_citations = create_cited_html(opinion, citations)
+
+        # Nuke existing citations
+        OpinionsCited.objects.filter(citing_opinion_id=opinion.pk).delete()
+
+        # Create the new ones.
+        OpinionsCited.objects.bulk_create(
+            [
+                OpinionsCited(
+                    citing_opinion_id=opinion.pk,
+                    cited_opinion_id=matched_opinion.pk,
+                    depth=grouped_matches[matched_opinion],
+                )
+                for matched_opinion in grouped_matches
+            ]
+        )
+
         # Update Solr if requested. In some cases we do it at the end for
         # performance reasons.
         opinion.save(index=index)
```

Comment:
Apropos of minimizing Solr updates, would it make sense to just do a simple `opinion.save()` call here, and then put another async `add_items_to_solr()` call outside of the for loop to to make Solr update the current batch of opinions all at once? I have no idea if this would improve performance (since we still need to do `opinion.save()` at least) -- just something that occurred to me.

(Note that this Solr update refers to updating the citing opinions; the one we already refactored above deals with updating the cited opinions.)

Reply:
Addressed in c5eefc5